File: preflight/health_check_orchestrator.py
"""ASTRA health check orchestrator - master coordinator of all preflight analysers."""
from __future__ import annotations

import time
import uuid
from typing import Any, Dict, Optional

from playwright.sync_api import sync_playwright

from preflight.anti_bot_analyser import AntiBotAnalyser
from preflight.bearer_token_analyser import BearerTokenAnalyser
from preflight.dom_analyser import DomAnalyser
from preflight.network_interceptor_analyser import NetworkInterceptorAnalyser
from utils.env_loader import ENV
from utils.logger import logger
from utils.report_generator import (
    AnalyserResult,
    PreflightReport,
    build_report,
    generate_reports,
)


def run_health_check() -> PreflightReport:
    """Master orchestrator: runs all 4 preflight analysers."""
    run_id = str(uuid.uuid4())[:8]
    start_time = time.time()

    logger.divider()
    logger.preflight(f"ASTRA Health Check [ID: {run_id}]")
    logger.preflight(f"Environment: {ENV.BASE_URL}")
    logger.divider()

    # Check if we should run health check
    if ENV.HEALTH_CHECK.lower() != "true":
        logger.preflight("Health check disabled (HEALTH_CHECK != true)")
        return PreflightReport(
            run_id=run_id,
            timestamp="",
            overall_status="SKIP",
            total_duration_ms=0,
            environment=ENV.BASE_URL,
            base_url=ENV.BASE_URL,
        )

    analysers: list[AnalyserResult] = []

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=ENV.HEADLESS.lower() == "true")
        context = browser.new_context()
        page = context.new_page()

        try:
            # 1. DOM Analyser
            dom_start = time.time()
            try:
                page.goto(ENV.TARGET_PAGE_URL)
                page.wait_for_load_state("networkidle")
                dom_result = DomAnalyser(page).analyse()
                analysers.append(AnalyserResult(
                    name="DOM Analyser",
                    status="PASS",
                    duration_ms=(time.time() - dom_start) * 1000,
                    findings=dom_result,
                ))
                logger.preflight_result("DOM Analyser", "OK")
            except Exception as exc:
                analysers.append(AnalyserResult(
                    name="DOM Analyser",
                    status="FAIL",
                    duration_ms=(time.time() - dom_start) * 1000,
                    errors=[str(exc)],
                ))
                logger.preflight_result("DOM Analyser", "FAIL", str(exc))

            # 2. Bearer Token Analyser
            token_start = time.time()
            try:
                token_result = BearerTokenAnalyser(page).analyse()
                analysers.append(AnalyserResult(
                    name="Bearer Token Analyser",
                    status="PASS" if token_result["token"] else "SKIP",
                    duration_ms=(time.time() - token_start) * 1000,
                    findings=token_result,
                ))
                logger.preflight_result("Bearer Token Analyser", token_result["status"])
            except Exception as exc:
                analysers.append(AnalyserResult(
                    name="Bearer Token Analyser",
                    status="FAIL",
                    duration_ms=(time.time() - token_start) * 1000,
                    errors=[str(exc)],
                ))
                logger.preflight_result("Bearer Token Analyser", "FAIL", str(exc))

            # 3. Network Interceptor Analyser
            network_start = time.time()
            try:
                network_result = NetworkInterceptorAnalyser(page).analyse()
                analysers.append(AnalyserResult(
                    name="Network Interceptor",
                    status="PASS",
                    duration_ms=(time.time() - network_start) * 1000,
                    findings=network_result,
                ))
                logger.preflight_result("Network Interceptor", "OK")
            except Exception as exc:
                analysers.append(AnalyserResult(
                    name="Network Interceptor",
                    status="FAIL",
                    duration_ms=(time.time() - network_start) * 1000,
                    errors=[str(exc)],
                ))
                logger.preflight_result("Network Interceptor", "FAIL", str(exc))

            # 4. Anti-Bot Analyser
            bot_start = time.time()
            try:
                bot_result = AntiBotAnalyser(page).analyse()
                analysers.append(AnalyserResult(
                    name="Anti-Bot Analyser",
                    status="PASS" if bot_result.is_testable else "WARN",
                    duration_ms=(time.time() - bot_start) * 1000,
                    findings={
                        "threat_level": bot_result.overall_threat_level,
                        "is_testable": bot_result.is_testable,
                        "recommendation": bot_result.recommendation,
                    },
                    warnings=[
                        f"{f.threat_type}: {f.threat_level}" for f in bot_result.findings
                    ],
                ))
                logger.preflight_result("Anti-Bot Analyser", "OK", bot_result.overall_threat_level)
            except Exception as exc:
                analysers.append(AnalyserResult(
                    name="Anti-Bot Analyser",
                    status="FAIL",
                    duration_ms=(time.time() - bot_start) * 1000,
                    errors=[str(exc)],
                ))
                logger.preflight_result("Anti-Bot Analyser", "FAIL", str(exc))

        finally:
            page.close()
            context.close()
            browser.close()

    total_duration = (time.time() - start_time) * 1000

    report = build_report(
        run_id=run_id,
        base_url=ENV.BASE_URL,
        environment="production",
        analysers=analysers,
        total_duration_ms=total_duration,
        metadata={"python_framework": "playwright", "version": "1.0.0"},
    )

    logger.divider()
    logger.preflight(f"Health Check Complete: {report.overall_status}")
    logger.preflight(f"Duration: {total_duration:.0f}ms")
    logger.divider()

    generate_reports(report)
    return report

File: preflight/healthCheck.orchestrator.ts
// ═══════════════════════════════════════════════════════════════
// preflight/healthCheck.orchestrator.ts
// ASTRA Framework — Health Check Orchestrator
// Runs all 4 analysers in sequence → generates preflight report
// Acts as Playwright globalSetup entry point
//Key highlights:
//
// safeRun() — error boundary wrapper around each analyser — one analyser crash never kills the whole preflight run
// shouldHardStop() — clean single-responsibility check: FAIL + HEALTH_CHECK=TRUE = stop, FAIL + HEALTH_CHECK=FALSE = warn + continue
// handleHardStop() — generates partial report even on hard stop — you always know what failed and why
// Summary table printed to console with box-drawing characters — clean visual at end of preflight run:
//
//   ┌─────────────────────────────────────┬──────────┬───────────┐
//   │ DOM Analyser                        │ ✅ PASS  │    842ms  │
//   │ Bearer Token Analyser               │ ✅ PASS  │   1203ms  │
//   │ Network Interceptor Analyser        │ ⚠️  WARN │   2100ms  │
//   │ Anti-Bot Analyser                   │ ✅ PASS  │    654ms  │
//   └─────────────────────────────────────┴──────────┴───────────┘
//
// export default runHealthCheck — Playwright globalSetup compatible
// require.main === module — also works as standalone npm run preflight
// Purple ASTRA banner printed at startup 🎨
// ═══════════════════════════════════════════════════════════════

import { validateEnv, ENV }              from "../utils/envLoader";
import { logger }                        from "../utils/logger";
import { generateReports, AnalyserResult } from "../utils/reportGenerator";
import { runDomAnalyser }                from "./domAnalyser";
import { runBearerTokenAnalyser }        from "./bearerTokenAnalyser";
import { runNetworkInterceptorAnalyser } from "./networkInterceptorAnalyser";
import { runAntiBotAnalyser }            from "./antiBotAnalyser";

// ═══════════════════════════════════════════════════════════════
// Orchestrator Entry Point
// Called by Playwright globalSetup + npm run preflight
// ═══════════════════════════════════════════════════════════════
async function runHealthCheck(): Promise<void> {
  // ─── Print ASTRA Banner ──────────────────────────────────────
  printBanner();

  // ─── Validate .env first ─────────────────────────────────────
  logger.divider("Environment Validation");
  const envResult = validateEnv();

  if (!envResult.valid) {
    logger.error("❌ Environment validation failed — missing required keys:");
    envResult.missingKeys.forEach((k) => logger.error(`   → ${k}`));
    logger.error("Please update your .env file and retry.");
    process.exit(1);
  }

  // Log warnings for optional keys using defaults
  envResult.warnings.forEach((w) => logger.warn(w));
  logger.info(`✅ Environment validated — ${envResult.loadedKeys.length} keys loaded`);
  logger.info(`🔧 Health Check Mode: ${ENV.HEALTH_CHECK ? "HARD STOP" : "WARN & PROCEED"}`);

  // ─── Run Analysers ───────────────────────────────────────────
  logger.divider("ASTRA Preflight Health Check");
  logger.info(`Target URL  : ${ENV.TARGET_PAGE_URL}`);
  logger.info(`Base URL    : ${ENV.BASE_URL}`);
  logger.info(`Browser     : ${ENV.BROWSER} | Headless: ${ENV.HEADLESS}`);

  const results: AnalyserResult[] = [];

  // ── 1. DOM Analyser ──────────────────────────────────────────
  const domResult = await safeRun("DOM Analyser", runDomAnalyser);
  results.push(domResult);
  if (shouldHardStop(domResult)) return handleHardStop(results);

  // ── 2. Bearer Token Analyser ─────────────────────────────────
  const tokenResult = await safeRun("Bearer Token Analyser", runBearerTokenAnalyser);
  results.push(tokenResult);
  if (shouldHardStop(tokenResult)) return handleHardStop(results);

  // ── 3. Network Interceptor Analyser ──────────────────────────
  const networkResult = await safeRun("Network Interceptor Analyser", runNetworkInterceptorAnalyser);
  results.push(networkResult);
  if (shouldHardStop(networkResult)) return handleHardStop(results);

  // ── 4. Anti-Bot Analyser ─────────────────────────────────────
  const antiBotResult = await safeRun("Anti-Bot Analyser", runAntiBotAnalyser);
  results.push(antiBotResult);
  if (shouldHardStop(antiBotResult)) return handleHardStop(results);

  // ─── Generate Final Report ───────────────────────────────────
  await finalizeReport(results);
}

// ═══════════════════════════════════════════════════════════════
// safeRun — wraps each analyser with error boundary
// Ensures one analyser crash doesn't kill the orchestrator
// ═══════════════════════════════════════════════════════════════
async function safeRun(
  name:    string,
  runner:  () => Promise<AnalyserResult>
): Promise<AnalyserResult> {
  try {
    logger.info(`\n⟳  Running: ${name}`);
    const result = await runner();
    return result;
  } catch (error) {
    const errMsg = error instanceof Error ? error.message : String(error);
    logger.error(`💥 ${name} crashed unexpectedly: ${errMsg}`);
    return {
      name,
      status:    "FAIL",
      duration:  0,
      details:   `Unexpected crash: ${errMsg}`,
      findings:  { error: errMsg },
      timestamp: new Date().toISOString(),
    };
  }
}

// ═══════════════════════════════════════════════════════════════
// shouldHardStop — checks HEALTH_CHECK flag + analyser status
// ═══════════════════════════════════════════════════════════════
function shouldHardStop(result: AnalyserResult): boolean {
  if (result.status !== "FAIL") return false;
  if (!ENV.HEALTH_CHECK) {
    logger.warn(
      `⚠️  [${result.name}] FAILED — HEALTH_CHECK=FALSE → Warning only, continuing...`
    );
    return false;
  }
  return true;
}

// ═══════════════════════════════════════════════════════════════
// handleHardStop — generates partial report + exits process
// ═══════════════════════════════════════════════════════════════
async function handleHardStop(results: AnalyserResult[]): Promise<void> {
  logger.divider("HARD STOP — Health Check Failed");
  logger.error("❌ ASTRA Preflight HARD STOP triggered");
  logger.error(`   Failed analyser: ${results[results.length - 1].name}`);
  logger.error(`   Reason: ${results[results.length - 1].details}`);
  logger.error("   Fix the issue and re-run preflight before proceeding.");

  // Still generate report with partial results
  await generateReports(results);

  logger.error("\n📄 Partial preflight report generated — check reports/preflight/");
  process.exit(1);
}

// ═══════════════════════════════════════════════════════════════
// finalizeReport — generates full report + prints summary table
// ═══════════════════════════════════════════════════════════════
async function finalizeReport(results: AnalyserResult[]): Promise<void> {
  logger.divider("Preflight Summary");

  // ─── Print summary table ─────────────────────────────────────
  const statusIcon: Record<string, string> = {
    PASS: "✅",
    FAIL: "❌",
    WARN: "⚠️ ",
    SKIP: "⏭️ ",
  };

  console.log("\n");
  console.log("  ┌─────────────────────────────────────┬──────────┬──────────┐");
  console.log("  │ Analyser                            │ Status   │ Duration │");
  console.log("  ├─────────────────────────────────────┼──────────┼──────────┤");

  for (const r of results) {
    const icon     = statusIcon[r.status] ?? "  ";
    const namePad  = r.name.padEnd(35);
    const statPad  = `${icon} ${r.status}`.padEnd(8);
    const durPad   = `${r.duration}ms`.padStart(8);
    console.log(`  │ ${namePad} │ ${statPad} │ ${durPad} │`);
  }

  console.log("  └─────────────────────────────────────┴──────────┴──────────┘");

  const passed  = results.filter((r) => r.status === "PASS").length;
  const failed  = results.filter((r) => r.status === "FAIL").length;
  const warned  = results.filter((r) => r.status === "WARN").length;
  const skipped = results.filter((r) => r.status === "SKIP").length;

  console.log(`\n  Total: ${results.length} | ✅ ${passed} | ❌ ${failed} | ⚠️  ${warned} | ⏭️  ${skipped}`);

  // ─── Generate JSON + HTML reports ────────────────────────────
  const report = await generateReports(results);

  // ─── Final gate decision ──────────────────────────────────────
  console.log("\n");
  if (report.overallPass) {
    logger.info("🚀 ASTRA Preflight PASSED — Proceeding to test execution");
  } else if (!ENV.HEALTH_CHECK) {
    logger.warn("⚠️  ASTRA Preflight completed with failures — HEALTH_CHECK=FALSE → Proceeding anyway");
  } else {
    logger.error("❌ ASTRA Preflight FAILED — Test execution blocked");
    process.exit(1);
  }

  logger.info(`\n📄 Reports saved to: ${ENV.REPORT_OUTPUT_DIR}/`);
  logger.info(`   → preflightReport.json`);
  logger.info(`   → preflightReport.html`);
}

// ═══════════════════════════════════════════════════════════════
// printBanner — ASTRA startup banner
// ═══════════════════════════════════════════════════════════════
function printBanner(): void {
  console.log(`
\x1b[35m
  ╔═══════════════════════════════════════════════════════╗
  ║                                                       ║
  ║        ✦  A S T R A  F R A M E W O R K  ✦            ║
  ║                                                       ║
  ║   Autonomous A* Search Based Test &                   ║
  ║   Reporting Architecture                              ║
  ║                                                       ║
  ║   Version : 1.0.0                                     ║
  ║   Mode    : Preflight Health Check                    ║
  ║                                                       ║
  ╚═══════════════════════════════════════════════════════╝
\x1b[0m`);
}

// ═══════════════════════════════════════════════════════════════
// Export for Playwright globalSetup
// ═══════════════════════════════════════════════════════════════
export default runHealthCheck;

// ═══════════════════════════════════════════════════════════════
// Direct execution: npm run preflight
// ═══════════════════════════════════════════════════════════════
if (require.main === module) {
  runHealthCheck().catch((error) => {
    logger.error(`ASTRA Preflight crashed: ${error}`);
    process.exit(1);
  });
}